[enterprise-4.15] TELCODOCS-1976: Adding content for improved SR-IOV Network Operator node draining (#79757)
openshift/openshift-docs: modules/nw-sriov-configuring-multiple-nodes.adoc

:_mod-docs-content-type: PROCEDURE
[id="configure-sr-iov-operator-parallel-nodes_{context}"]
= Configuring parallel node draining during SR-IOV network policy updates

By default, the SR-IOV Network Operator drains workloads from a node before every policy change.

> The Operator performs this action, one node at a time, to ensure that no workloads are affected by the reconfiguration.

Reviewer: ```suggestion
The Operator completes this action, one node at a time, to ensure that no workloads are affected by the reconfiguration.
```

In large clusters, draining nodes sequentially can be time-consuming, taking hours or even days. In time-sensitive environments, you can enable parallel node draining in an `SriovNetworkPoolConfig` custom resource (CR) for faster rollouts of SR-IOV network configurations. 

To configure parallel draining, use the `SriovNetworkPoolConfig` CR to create a node pool. You can then add nodes to the pool and define the maximum number of nodes in the pool that the Operator can drain in parallel. With this approach, you can enable parallel draining for faster reconfiguration while ensuring you still have enough nodes remaining in the pool to handle any running workloads.

[NOTE]
====

> A node can only belong to one SR-IOV network pool configuration. If a node is not part of a pool, it is added to a virtual, default, pool that is configured to drain one node at a time only.

Reviewer: ```suggestion
A node can belong to only one SR-IOV network pool configuration. If a node is not part of a pool, it is added to a virtual, default pool that is configured to drain one node at a time only.
```

The node might restart during the draining process.
====

.Prerequisites

* Install the OpenShift CLI (`oc`).
* Log in as a user with `cluster-admin` privileges.
* Install the SR-IOV Network Operator.

> * Nodes have hardware that support SR-IOV.

Reviewer: ```suggestion
* Ensure that nodes have hardware that supports SR-IOV.
```

.Procedure

. Create a `SriovNetworkPoolConfig` resource:

.. Create a YAML file that defines the `SriovNetworkPoolConfig` resource:
+
.Example `sriov-nw-pool.yaml` file
[source,yaml]
----
apiVersion: v1
kind: SriovNetworkPoolConfig
metadata:
  name: pool-1 <1>
  namespace: openshift-sriov-network-operator <2>
spec:
  maxUnavailable: 2 <3>
  nodeSelector: <4>
    matchLabels:
      node-role.kubernetes.io/worker: ""
----
<1> Specify the name of the `SriovNetworkPoolConfig` object.
<2> Specify namespace where the SR-IOV Network Operator is installed.

> <3> Specify an integer number, or percentage value, for nodes that can be unavailable in the pool during an update. For example, if you have 10 nodes and you set the maximum unavailable to 2, then only 2 nodes can be drained in parallel at any time, leaving 8 nodes for handling workloads.

Reviewer: ```suggestion
<3> Specify an integer number or percentage value for nodes that can be unavailable in the pool during an update. For example, if you have 10 nodes and you set the maximum unavailable value to 2, then only 2 nodes can be drained in parallel at any time, leaving 8 nodes for handling workloads.
```

  nodeSelector:
    node-role.kubernetes.io/worker: ""
  numVfs: 5
  priority: 99
  resourceName: sriov_nic_1
----

.. Create the `SriovNetworkNodePolicy` resource by running the following command:
+
[source,terminal]
----
$ oc create -f sriov-node-policy.yaml
----

. Create a `SriovNetwork` resource:

.. Create a YAML file that defines the `SriovNetwork` resource:
+
.Example `sriov-network.yaml` file
[source,yaml]
----
apiVersion: sriovnetwork.openshift.io/v1
kind: SriovNetwork
metadata:
  name: sriov-nic-1
  namespace: openshift-sriov-network-operator
spec:
  linkState: auto
  networkNamespace: sriov-test
  resourceName: sriov_nic_1
  capabilities: '{ "mac": true, "ips": true }'
  ipam: '{ "type": "static" }'
----

.. Create the `SriovNetwork` resource by running the following command:
+
[source,terminal]
----
$ oc create -f sriov-network.yaml
----

.Verification 

* View the node pool you created by running the following command:
+
[source,terminal]
----
$ oc get sriovNetworkpoolConfig -n openshift-sriov-network-operator
----
+
.Example output
[source,terminal]
----
NAME     AGE
pool-1   67s <1>
----
<1> In this example, `pool-1` contains all the nodes with the `worker` role.

> To demonstrate the node draining process using the example scenario from the above procedure, complete the following steps:

Reviewer: ```suggestion
To demonstrate the node draining process by using the example scenario from the previous procedure, complete the following steps:
```
Should this be formatted as a bulleted item like the bulleted item above it on line 124?
